src/adsp_config.c

```diff
@@ -1575,35 +1589,21 @@ static int adsp_parse_config_fd(FILE *fd, struct adsp *out, bool verbose)
 	/* manifest version is in toml root */
 	ret = parse_version(toml, manifest_version);
 	if (ret < 0)
-		goto error;
-	}
+		goto out;
 
-	/* parse "version" array elements */
-	for (i = 0; i < ARRAY_SIZE(manifest_version); ++i) {
-		raw = toml_raw_at(arr, i);
-		if (ret < 0)
-			ret = err_key_parse("version", NULL);
-		ret = toml_rtoi(raw, &manifest_version[i]);
-		if (ret < 0)
-			ret = err_key_parse("version", "can't convert element to integer");
+	/* look for parser related with manifest version and run it */
+	for (i = 0; i < ARRAY_SIZE(parsers); ++i) {
+		if (parsers[i].major == manifest_version[0] &&
+		    parsers[i].minor == manifest_version[1]) {
+			ret = parsers[i].parse(toml, out, verbose);
+			goto out;
+		}
 	}
 
-	/* parsing function depends on manifest_version */
-	if (check_config_version(1, 0, manifest_version)) {
-		ret = parse_adsp_config_v1_0(toml, out, verbose);
-	} else if (check_config_version(1, 5, manifest_version)) {
-		ret = parse_adsp_config_v1_5(toml, out, verbose);
-	} else if (check_config_version(1, 8, manifest_version)) {
-		ret = parse_adsp_config_v1_8(toml, out, verbose);
-	} else if (check_config_version(2, 5, manifest_version)) {
-		ret = parse_adsp_config_v2_5(toml, out, verbose);
-	} else {
-		ret = log_err(-EINVAL, "error: Unsupported config version %d.%d\n",
-			      manifest_version[0], manifest_version[1]);
-		goto error;
-	}
+	ret = log_err(-EINVAL, "error: Unsupported config version %d.%d\n",
+		      manifest_version[0], manifest_version[1]);
```

Comment:
This really obfuscates the `ret` value to save extremely little code.

Reply:
Ok, so I extracted `find_config_parser()` function, then error flow is everywhere inside `if` statement and success flow is outside `if`